src/map.mo

```diff
@@ -222,48 +210,39 @@ module {
           x := x * 256 + Nat64.fromNat(Nat8.toNat(bytes[Int.abs(i)]));
         };
         if (x > 0) {
-          if (lastNode != 0) return #multiple;
+          if (lastNode != 0) return node;
           lastNode := x;
         };
       };
-      if (lastNode == 0) return #empty;
-      if (lastNode & 1 == 0) return #multiple;
-      return #single lastNode;
+      if (lastNode & 1 == 0) node else lastNode;
     };
 
-    func deleteRec(nodes : Base.Region, leaves : Base.Region, key : Blob, bytes : [Nat8], node : Nat64, pos : Nat16) : (?Blob, DeleteRes) {
-      if (node == 0) return (null, #multiple);
+    func deleteRec(nodes : Base.Region, leaves : Base.Region, key : Blob, bytes : [Nat8], node : Nat64, pos : Nat16) : (?Blob, Nat64) {
+      if (node == 0) return (null, node);
       if (node & 1 == 1) {
         let leaf = node >> 1;
         if (base.getKey(leaves, leaf) == key) {
-          let ret = (?base.getValue(leaves, leaf), #empty);
+          let ret = (?base.getValue(leaves, leaf), 0 : Nat64);
           pushEmptyLeaf(leaves, leaf);
           return ret;
         } else {
-          return (null, #multiple)
-        }
+          return (null, node);
+        };
       };
 
       let idx = base.keyToIndex(bytes, pos);
       let child = base.getChild(nodes, node, idx);
       let ret = deleteRec(nodes, leaves, key, bytes, child, pos +% base.bitlength);
-      let (value, leaf_opt) = ret;
+      let (value, leaf) = ret;
 
-      let ret_leaf = switch (leaf_opt) {
-        case (#empty) {
-          base.setChild(nodes, node, idx, 0);
-          singleLeaf(nodes, node);
-        };
-        case (#multiple) {
-          #multiple;
-        };
-        case (#single(leaf)) {
-          base.setChild(nodes, node, idx, leaf);
-          singleLeaf(nodes, node);
-        };
+      let ret_leaf = if (leaf != child) {
+        base.setChild(nodes, node, idx, leaf);
+        singleLeaf(nodes, node);
```

Comment:
We should test if `leaf` (aka `branch_root`) is an inner node or not. If it is an inner node we can skip the call to `singleLeaf`.